[CIVP-12555] fix jupyter widgets

mheilman opened the pull request:
The jupyter widgets package (ipywidgets) was included in the image previously via the jupyter dependency, but they didn't work.

This pins the `ipywidgets` and `notebook` versions and adds the nice `bqplot` viz package that uses jupyter widgets in `environment.yml`.  It also adds a command to the dockerfile to enable jupyter widgets since the conda package doesn't appear to do that.

mheilman committed:
this makes sure that jupyter widgets are appropriately enabled in the image (they did not work before). also adds bqplot

mheilman committed:
change expected conda forge package count (default channel appears to have more packages now)

Also updates the changelog.

3.3.0 output:
> conda list|grep  conda-forge
arrow-cpp                 0.7.1                    py36_2    conda-forge
backports.functools_lru_cache 1.4                      py36_1    conda-forge
fontconfig                2.12.1                        5    conda-forge
freetype                  2.7                           2    conda-forge
matplotlib                2.1.0                    py36_1    conda-forge
parquet-cpp               1.3.0.post                    2    conda-forge
pyarrow                   0.7.1                    py36_1    conda-forge
qt                        5.6.2                         6    conda-forge
xgboost                   0.6a2                    py36_2    conda-forge

current output:

> conda list |grep conda-forge
arrow-cpp                 0.7.1                    py36_2    conda-forge
bqplot                    0.10.1                   py36_0    conda-forge
fontconfig                2.12.6                        0    conda-forge
freetype                  2.8.1                         0    conda-forge
parquet-cpp               1.3.0.post                    2    conda-forge
pyarrow                   0.7.1                    py36_1    conda-forge
traittypes                0.0.6                    py36_0    conda-forge
xgboost                   0.6a2                    py36_2    conda-forge

mheilman commented:
@stephen-hoover, no rush

stephen-hoover commented on Dockerfile:
```diff
@@ -92,6 +92,11 @@ RUN mkdir -p ${HOME}/.config/matplotlib && \
+# Enable widgetsnbextension for jupyter widgets.
+# See https://ipywidgets.readthedocs.io/en/stable/user_install.html.
+# This enables the extension in the conda environment.
+RUN jupyter nbextension enable --py widgetsnbextension
```
Is this still necessary if you explicitly install the ipywidgets package, instead of letting it be pulled in as a dependency? The documentation you link to says this command shouldn't be needed when installing from conda.

beckermr replied:
Random comment: I have also found that you do not need the command above when installing from conda locally. Conda has a post-link feature which allows for stuff like this after packages are installed.

mheilman committed:
update ipywidgets and bqplot versions, remove unneeded widgetsnbextension enable call

Also update conda-forge count to 10:

> conda list|grep  conda-forge
arrow-cpp                 0.7.1                    py36_2    conda-forge
bqplot                    0.10.2                   py36_0    conda-forge
fontconfig                2.12.6                        0    conda-forge
freetype                  2.8.1                         0    conda-forge
ipywidgets                7.1.0                    py36_0    conda-forge
parquet-cpp               1.3.0.post                    2    conda-forge
pyarrow                   0.7.1                    py36_1    conda-forge
traittypes                0.0.6                    py36_0    conda-forge
widgetsnbextension        3.1.0                    py36_0    conda-forge
xgboost                   0.6a2                    py36_2    conda-forge

mheilman commented:
I updated ipywidgets to 7.1.0, which wasn't out when I made this PR originally, and now I don't need the `jupyter nbextension enable` call.  I'm pretty sure this has to do with installing from conda-forge vs the conda default channel, which, I believe doesn't enable the necessary extension after installing it ([see related comment here](https://github.com/conda-forge/ipywidgets-feedstock/issues/31#issuecomment-348228172)).

Perhaps we should have conda-forge be at the top of the list of channels in `environment.yml`.  Or maybe we should be explicit about where packages come from ([like this?](https://conda.io/docs/user-guide/tasks/manage-environments.html#building-identical-conda-environments)).

mheilman committed:
updated versions in changelog

beckermr commented:
Yeah. Conda forge seems to be getting better fwiw. They now ship proper binaries for openmp on macs for example. 

They also have a great MYSQL package. ;)

Might be time to make the leap.

stephen-hoover commented:
I could be convinced to pull from conda-forge first and Anaconda second, but this wouldn't be the PR to do it. Hm. Now that you're pointing out that only the conda-forge install automatically runs `jupyter nbextension enable --py widgetsnbextension` for you, I'm wondering if it's safer to keep it in our Dockerfile. Until / unless we pull from conda-forge first, the widgets will break again when Anaconda publishes v7.1.0 binaries.

beckermr commented:
Ack yes @stephen-hoover.

mheilman committed:
add back nbextension enable command just in case

mheilman commented:
OK, I added the `enable` command back in, this time with a comment about the difference between channels.

stephen-hoover reviewed (approved):
LGTM! Thanks for the clarification about when the `nbextension enable` is needed.
